File: scripts/setup_claude_mcp.js
#!/usr/bin/env node
"use strict";

const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");

const SERVER_NAME = "xmux_bridge";
const LEGACY_NAMES = new Set([
  "xmux_bridge",
  "xmux-bridge",
  "clau_mux_bridge",
  "clau-mux-bridge",
  "amux_bridge",
  "amux-bridge",
]);

function absolute(inputPath) {
  return path.resolve(inputPath.replace(/^~(?=$|\/)/, os.homedir()));
}

function stableHomebrewXmuxInstallDir(installDir) {
  const resolved = absolute(installDir);
  const marker = `${path.sep}Cellar${path.sep}xmux${path.sep}`;
  if (!resolved.includes(marker) || !resolved.endsWith(`${path.sep}libexec`)) {
    return resolved;
  }

  const prefix = resolved.split(marker, 1)[0];
  const candidate = path.join(prefix, "opt", "xmux", "libexec");
  if (fs.existsSync(path.join(candidate, "xmux.zsh"))) {
    return candidate;
  }
  return resolved;
}

function stableHomebrewXmuxFilePath(filePath) {
  const resolved = absolute(filePath);
  const installDir = path.dirname(resolved);
  const stableInstallDir = stableHomebrewXmuxInstallDir(installDir);
  if (stableInstallDir === installDir) {
    return resolved;
  }

  const candidate = path.join(stableInstallDir, path.basename(resolved));
  if (fs.existsSync(candidate)) {
    return candidate;
  }
  return resolved;
}

function loadJson(filePath) {
  if (!fs.existsSync(filePath)) {
    return {};
  }
  const parsed = JSON.parse(fs.readFileSync(filePath, "utf-8"));
  if (parsed === null || Array.isArray(parsed) || typeof parsed !== "object") {
    throw new Error(`error: ${filePath} must contain a JSON object`);
  }
  return parsed;
}

function atomicWriteJson(filePath, data) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  const tmpPath = path.join(
    path.dirname(filePath),
    `.${path.basename(filePath)}.${process.pid}.${Date.now()}.tmp`,
  );
  fs.writeFileSync(tmpPath, `${JSON.stringify(data, null, 2)}\n`, "utf-8");
  fs.renameSync(tmpPath, filePath);
}

function usage() {
  process.stderr.write(
    "usage: setup_claude_mcp.js <bridge_js> <project_dir> <outbox> <agent> <team> <state_dir> <install_dir>\n",
  );
}

function main(argv = process.argv.slice(2)) {
  if (argv.length !== 7) {
    usage();
    return 2;
  }

  const bridgeJs = stableHomebrewXmuxFilePath(argv[0]);
  const projectDir = absolute(argv[1]);
  const outbox = absolute(argv[2]);
  const agent = argv[3];
  const team = argv[4];
  const stateDir = absolute(argv[5]);
  const installDir = stableHomebrewXmuxInstallDir(argv[6]);

  const configPath = path.join(os.homedir(), ".claude.json");
  const config = loadJson(configPath);

  const projects = config.projects ?? {};
  if (projects === null || Array.isArray(projects) || typeof projects !== "object") {
    throw new Error("error: ~/.claude.json projects must be a JSON object");
  }
  config.projects = projects;

  const project = projects[projectDir] ?? {};
  if (project === null || Array.isArray(project) || typeof project !== "object") {
    throw new Error(`error: Claude project entry for ${projectDir} must be a JSON object`);
  }
  projects[projectDir] = project;

  const servers = project.mcpServers ?? {};
  if (servers === null || Array.isArray(servers) || typeof servers !== "object") {
    throw new Error(`error: Claude mcpServers for ${projectDir} must be a JSON object`);
  }
  project.mcpServers = servers;

  for (const legacyName of LEGACY_NAMES) {
    delete servers[legacyName];
  }

  servers[SERVER_NAME] = {
    type: "stdio",
    command: "node",
    args: [bridgeJs, "--outbox", outbox, "--agent", agent, "--team", team],
    env: {
      XMUX_AGENT: agent,
      XMUX_INSTALL_DIR: installDir,
      XMUX_OUTBOX: outbox,
      XMUX_PROJECT_DIR: projectDir,
      XMUX_STATE_DIR: stateDir,
      XMUX_TEAM: team,
    },
  };

  atomicWriteJson(configPath, config);
  return 0;
}

if (require.main === module) {
  try {
    process.exitCode = main();
  } catch (error) {
    process.stderr.write(`${String(error.message || error)}\n`);
    process.exitCode = 1;
  }
}

module.exports = {
  main,
  stableHomebrewXmuxInstallDir,
  stableHomebrewXmuxFilePath,
};

File: archive/legacy-python/scripts/setup_claude_mcp.py
#!/usr/bin/env python3
"""Configure Claude Code local MCP for XMux teammate callbacks.

Usage:
  python3 setup_claude_mcp.py <bridge_js> <project_dir> <outbox> <agent> <team> <state_dir> <install_dir>
"""

from __future__ import annotations

import json
import os
import sys
import tempfile
from pathlib import Path

SERVER_NAME = "xmux_bridge"
LEGACY_NAMES = {
    "xmux_bridge",
    "xmux-bridge",
    "clau_mux_bridge",
    "clau-mux-bridge",
    ("a" + "mux_bridge"),
    ("a" + "mux-bridge"),
}


def _absolute(path: str) -> str:
    return os.path.abspath(os.path.expanduser(path))


def _stable_homebrew_xmux_install_dir(install_dir: str) -> str:
    resolved = _absolute(install_dir)
    marker = f"{os.sep}Cellar{os.sep}xmux{os.sep}"
    if marker not in resolved or not resolved.endswith(f"{os.sep}libexec"):
        return resolved

    prefix = resolved.split(marker, 1)[0]
    candidate = os.path.join(prefix, "opt", "xmux", "libexec")
    if os.path.isfile(os.path.join(candidate, "xmux.zsh")):
        return candidate
    return resolved


def _stable_homebrew_xmux_file_path(path: str) -> str:
    resolved = _absolute(path)
    install_dir = os.path.dirname(resolved)
    stable_install_dir = _stable_homebrew_xmux_install_dir(install_dir)
    if stable_install_dir == install_dir:
        return resolved

    candidate = os.path.join(stable_install_dir, os.path.basename(resolved))
    if os.path.isfile(candidate):
        return candidate
    return resolved


def _load_json(path: Path) -> dict:
    if not path.exists():
        return {}
    with path.open(encoding="utf-8") as fh:
        data = json.load(fh)
    if not isinstance(data, dict):
        raise SystemExit(f"error: {path} must contain a JSON object")
    return data


def _atomic_write_json(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(
        prefix=f".{path.name}.",
        suffix=".tmp",
        dir=str(path.parent),
        text=True,
    )
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            json.dump(data, fh, indent=2, sort_keys=True, ensure_ascii=True)
            fh.write("\n")
        os.replace(tmp_name, path)
    except Exception:
        try:
            os.unlink(tmp_name)
        except FileNotFoundError:
            pass
        raise


def _usage() -> None:
    print(
        "usage: setup_claude_mcp.py <bridge_js> <project_dir> <outbox> "
        "<agent> <team> <state_dir> <install_dir>",
        file=sys.stderr,
    )


def main() -> int:
    if len(sys.argv) != 8:
        _usage()
        return 2

    bridge_js = _stable_homebrew_xmux_file_path(sys.argv[1])
    project_dir = _absolute(sys.argv[2])
    outbox = _absolute(sys.argv[3])
    agent = sys.argv[4]
    team = sys.argv[5]
    state_dir = _absolute(sys.argv[6])
    install_dir = _stable_homebrew_xmux_install_dir(sys.argv[7])

    config_path = Path(os.path.expanduser("~/.claude.json"))
    config = _load_json(config_path)

    projects = config.setdefault("projects", {})
    if not isinstance(projects, dict):
        raise SystemExit("error: ~/.claude.json projects must be a JSON object")

    project = projects.setdefault(project_dir, {})
    if not isinstance(project, dict):
        raise SystemExit(f"error: Claude project entry for {project_dir} must be a JSON object")

    servers = project.setdefault("mcpServers", {})
    if not isinstance(servers, dict):
        raise SystemExit(f"error: Claude mcpServers for {project_dir} must be a JSON object")

    for name in LEGACY_NAMES:
        servers.pop(name, None)

    servers[SERVER_NAME] = {
        "type": "stdio",
        "command": "node",
        "args": [
            bridge_js,
            "--outbox",
            outbox,
            "--agent",
            agent,
            "--team",
            team,
        ],
        "env": {
            "XMUX_AGENT": agent,
            "XMUX_INSTALL_DIR": install_dir,
            "XMUX_OUTBOX": outbox,
            "XMUX_PROJECT_DIR": project_dir,
            "XMUX_STATE_DIR": state_dir,
            "XMUX_TEAM": team,
        },
    }

    _atomic_write_json(config_path, config)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
